RubberduckTests/Rewriter/CodePaneRewriteSessionTests.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System;
+using System.Linq;
+using Moq;
+using NUnit.Framework;
+using Rubberduck.Parsing.Rewriter;
+using Rubberduck.Parsing.VBA;
+using Rubberduck.Parsing.VBA.Parsing;
+using Rubberduck.VBEditor;
+
+namespace RubberduckTests.Rewriter
+{
+    [TestFixture]
+    public class CodePaneRewriteSessionTests : RewriteSessionTestBase
+    {
+        [Test]
+        [Category("Rewriter")]
+        public void RequestsParseAfterRewriteIfNotInvalidatedAndParsingAllowed()
+        {
+            var mockParseManager = new Mock<IParseManager>();
+            mockParseManager.Setup(m => m.OnParseRequested(It.IsAny<object>()));
+
+            var rewriteSession = RewriteSession(mockParseManager.Object, session => true, out _);
+            var module = new QualifiedModuleName("TestProject", string.Empty, "TestModule");
+            var otherModule = new QualifiedModuleName("TestProject", string.Empty, "OtherTestModule");
+            rewriteSession.CheckOutModuleRewriter(module);
+            rewriteSession.CheckOutModuleRewriter(otherModule);
+
+            rewriteSession.Rewrite();
+
+            mockParseManager.Verify(m => m.OnParseRequested(It.IsAny<object>()), Times.Once);
+        }
+
+        [Test]
+        [Category("Rewriter")]
+        public void DoesNotCallRequestsParseAfterRewriteIfInvalidated()
+        {
+            var mockParseManager = new Mock<IParseManager>();
+            mockParseManager.Setup(m => m.OnParseRequested(It.IsAny<object>()));
+
+            var rewriteSession = RewriteSession(mockParseManager.Object, session => true, out _);
+            var module = new QualifiedModuleName("TestProject", string.Empty, "TestModule");
+            var otherModule = new QualifiedModuleName("TestProject", string.Empty, "OtherTestModule");
+            rewriteSession.CheckOutModuleRewriter(module);
+            rewriteSession.Invalidate();
+            rewriteSession.CheckOutModuleRewriter(otherModule);
+
+            rewriteSession.Rewrite();
+
+            mockParseManager.Verify(m => m.OnParseRequested(It.IsAny<object>()), Times.Never);
```

Comment:
Why would invalidating not allow parse request? We had a module checked out and rewritten after the invalidation; if that is not legal, those should be an error on the checking out, rather than a no-op for reparse?

Reply:
The actual code will log a warning entry about rewrite attempt on an invalidated session. This test ensures that no reparse is requested; the session does not apply any rewrites anymore.

Reply:
I would suggest that this makes for a confusing API if I am able to execute an `Rewrite()` and nothing is thrown back at me. Logging is good but the calling code has no idea that the rewriting did not happen and may not behave as expected.

Reply:
My problem with throwing an exception is that I will have to wrap every call to `Rewrite` in a try catch block because it might throw.

So far, most of RD's functionality simply bails out when it encounters a situation in which the action requested does not work, e.g. when a refactoring thinks the data is not sound, it simply does not do anything. 

That a rewrite fails basically means that somebody else was faster and your work is outdated now. To me, it does not feel right to throw an exception for that.

I could live with adding a `TryRewrite`. However, I would really prefer not to do anything that requires changing the about one hundred callers.

Reply:
I think I misremembered who is calling the actual rewrite. There are probably only about 10 callers for `Rewrite`. The hundred callers are for checking out rewriters.